### PR TITLE
Copter: Throw mode changes

### PR DESCRIPTION
**Pull request by Chambana**
Added Downward Throw (AKA "Air Drop") functionality to Throw Mode. Upward or downward throw is settable via THROW_TYPE param.
Added "Next Mode" functionality to Throw Mode, which defaults to GUIDED, but is settable by param THROW_NEXT_MODE

### CODE REVIEW

**Comment by magicrub** [comment]
I see a merge in your git commits. Can you see about removing that?

**Comment by magicrub** [comment]
this replaces PR https://github.com/ArduPilot/ardupilot/pull/4372

@Chambana I see this is your 3rd PR for this code change. Would you mind force-pushing changes to the same branch instead of closing and opening new PR's for further changes?

**Comment by Chambana** [comment]
removed merge

**Comment by magicrub** [comment]
The type should be an enum

File: ArduCopter/control_throw.cpp [hunk 1]
```diff
@@ -95,9 +95,14 @@ void Copter::throw_run()
         pos_control.set_speed_z(BRAKE_MODE_SPEED_Z, BRAKE_MODE_SPEED_Z);
         pos_control.set_accel_z(BRAKE_MODE_DECEL_RATE);
 
-        // initialise the demanded height to 3m above the throw height
-        // we want to rapidly clear surrounding obstacles
-        pos_control.set_alt_target(inertial_nav.get_altitude() + 300);
+        // initialise the demanded height to 3m above the throw height or 1 meter below the drop height
+         // we want to rapidly clear surrounding obstacles
+        if (g.throw_type==1){
```
Comment (magicrub) [inline ArduCopter/control_throw.cpp]: Needs a space between the bracket and this needs to be an enum.

**Comment by lucasdemarchi** [comment]
Commits are in bad shape but can be fixed while applying. I think we should check the changes and get and ack from @priseborough

File: ArduCopter/control_throw.cpp [hunk 2]
```diff
@@ -95,9 +95,14 @@ void Copter::throw_run()
         pos_control.set_speed_z(BRAKE_MODE_SPEED_Z, BRAKE_MODE_SPEED_Z);
         pos_control.set_accel_z(BRAKE_MODE_DECEL_RATE);
 
-        // initialise the demanded height to 3m above the throw height
-        // we want to rapidly clear surrounding obstacles
-        pos_control.set_alt_target(inertial_nav.get_altitude() + 300);
+        // initialise the demanded height to 3m above the throw height or 1 meter below the drop height
```
Comment (priseborough) [inline ArduCopter/control_throw.cpp]: If the intention is for this to be used for air drop, then 1m does not seem like enough margin given baro and other height estimation errors.

**Comment by priseborough** [comment]
The logic looks OK. I have concerns about arresting the height only 1m below the detection height, given height estimation errors, but maybe this is OK given the detection delay. Is there any test data (SITL or flight) demonstrating that both the normal throw and drop launch work after these patches are applied?  I would not be pushing changes to flight critical logic like this without test verification.

**Comment by Chambana** [comment]
@priseborough 1m was a bit arbitrary and, yes, I haven't been able to test it with a real aircraft.   If it needs to be tested prior to inclusion in master, I can try to figure out a way to test it.

**Comment by priseborough** [comment]
ArduPlane SITL had a ballon drop simulation capability, so maybe this functionality can be ported to Copter SITL.

**Comment by rmackay9** [comment]
This change has been incorporated into this new PR: https://github.com/ArduPilot/ardupilot/pull/4592

**Pull request closed without merging**